File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def deleteNode(self, root, key):
        """
        :type root: Optional[TreeNode]
        :type key: int
        :rtype: Optional[TreeNode]
        """
        if root == None:
            return
        pre = None
        curr = root
        whchDirection = None

        while curr is not None:
            if key > curr.val:
                pre = curr
                curr = curr.right
            elif key < curr.val:
                pre = curr
                curr = curr.left
            elif curr.val == key:
                if pre is not None and pre.right is not None and pre.right.val == key:                    
                    whchDirection = "right"
                else:
                    whchDirection = "left"
                break
        
        if curr == None:
            return root

        if pre is not None:
            if curr.right == None and curr.left == None:
                if whchDirection == "left":
                    pre.left = None
                else:
                    pre.right = None
            elif curr.right is not None and curr.left == None:
                if whchDirection == "left":
                    pre.left = curr.right
                else:
                    pre.right = curr.right
            elif curr.left is not None and curr.right == None:
                if whchDirection == "left":
                    pre.left = curr.left
                else:
                    pre.right = curr.left
            else:
                nxtCur = curr.right
                if nxtCur.left == None and nxtCur.right == None:
                    curr.val = nxtCur.val
                    curr.right = None
                elif nxtCur.left == None and nxtCur.right is not None:
                    curr.val = nxtCur.val
                    curr.right = nxtCur.right

                else:
                    prePre = None
                    while nxtCur.left is not None:
                        prePre = nxtCur
                        nxtCur = nxtCur.left
                    if nxtCur.right == None:
                        curr.val = nxtCur.val
                        prePre.left = None
                    else:
                        curr.val = nxtCur.val
                        prePre.left = nxtCur.right
        else:
            if curr.left == None and curr.right == None:
                root = None
            elif curr.right is not None and curr.left == None:
                root = root.right
            elif curr.left is not None and curr.right == None:
                root = root.left
            else:
                nxtCur = curr.right
                if nxtCur.left == None and nxtCur.right == None:
                    curr.val = nxtCur.val
                    curr.right = None
                elif nxtCur.left == None and nxtCur.right is not None:
                    curr.val = nxtCur.val
                    curr.right = nxtCur.right
                else:
                    prePre = None
                    while nxtCur.left is not None:
                        prePre = nxtCur
                        nxtCur = nxtCur.left
                    if nxtCur.right == None:
                        curr.val = nxtCur.val
                        prePre.left = None
                    else:
                        curr.val = nxtCur.val
                        prePre.left = nxtCur.right

        return root



        

File: 0450-delete-node-in-a-bst/0450-delete-node-in-a-bst.c
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

int findSmallestValueinBST(struct TreeNode* root);

struct TreeNode* deleteNode(struct TreeNode* root, int key) {


    if(!root) return root;
    if(root->val == key ){

        if(!(root->right) && !(root->left)){
            root = NULL;
            return root;
        }else if(!(root->right) && root->left){
            root = root->left;
            return root;
        } else if (!(root->left) && root->right){
            root = root->right;
            return root;
        }else{  
            int s = findSmallestValueinBST(root->right);
            root->right = deleteNode(root->right, s);
            root->val = s;
            return root;
        }

    } else {

        struct TreeNode* copy = root;
        struct TreeNode* pre = NULL;
        while(copy && copy->val != key){
            pre = copy;
            if(key > copy->val ){
                copy = copy->right;
            }else {
                copy = copy->left;
            }
        }


        if(!copy) return root;

        if(!(copy->left) && !(copy->right)){
            if(pre->left && pre->left->val == key) {
                pre->left = NULL;
            } else{
                pre->right = NULL;
            }
        }else if(!(copy->left) && copy->right){
            if(pre->left && pre->left->val == key) {
                pre->left = copy->right;
            } else{
                pre->right = copy->right;
            }
        }else if(!(copy->right) && copy->left){
            if(pre->left && pre->left->val == key) {
                pre->left = copy->left;
            } else{
                pre->right = copy->left;
            }
        }else{
            int smallestVal = findSmallestValueinBST(copy->right);
            copy->right = deleteNode(copy->right, smallestVal);
            copy->val = smallestVal;
        }

        return root;
    }

    
}

int findSmallestValueinBST(struct TreeNode* root){
    // int val = NULL;

    while(root->left){
        root = root->left;
    }

    return root->val;
}